Pytorch Fully Sharded Data Parallel (FSDP) Integration (#147)
ml-energy/zeus: examples/power_limit_optimizer/README.md

- [`ZeusMonitor`](http://ml.energy/zeus/reference/monitor/#zeus.monitor.ZeusMonitor): Measures the GPU time and energy consumption of arbitrary code blocks.
- [`GlobalPowerLimitOptimizer`](https://ml.energy/zeus/reference/optimizer/power_limit/#zeus.optimizer.power_limit.GlobalPowerLimitOptimizer): Online-profiles each power limit with `ZeusMonitor` and finds the cost-optimal power limit.

## Multi-GPU Distributed Training (Pytorch DDP and FSDP)

> When using `ZeusMonitor` and/or `GlobalPowerLimitOptimizer` in a multi-GPU Distributed context, launch one instance of `ZeusMonitor` and/or `GlobalPowerLimitOptimizer` per local rank (per GPU on each node), and pass in the local rank to `ZeusMonitor` as shown below:

Reviewer: ```suggestion
When using `ZeusMonitor` and/or `GlobalPowerLimitOptimizer` in a multi-GPU Distributed context, construct one instance of `ZeusMonitor` and/or `GlobalPowerLimitOptimizer` per local rank (per GPU on each node), and pass in the local rank to `ZeusMonitor` as shown below:
```

```python
monitor = ZeusMonitor(gpu_indices=[local_rank]) # pass in local rank to gpu_indices.
plo = GlobalPowerLimitOptimizer(monitor)
```

> Ensure that only one GPU is monitored per `ZeusMonitor`. Internally, `GlobalPowerLimitOptimizer` performs an [All-Reduce](https://pytorch.org/docs/stable/distributed.html) to synchronize before making a power limit decision.

Reviewer: ```suggestion
Ensure that only one GPU is monitored per `ZeusMonitor`. Internally, `GlobalPowerLimitOptimizer` performs an [AllReduce](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html) to aggregate time and energy measurements across all GPUs before making a power limit decision.
```

    [DATA_DIR] \
    --gpu 0                 `# Specify the GPU id to use`

# Multi-GPU Distributed Data Parallel
torchrun \
    --nnodes 1 \
    --nproc_per_node gpu    `# Number of processes per node, should be equal to the number of GPUs.` \
                            `# When set to 'gpu', it means use all the GPUs available.` \
    train_dp.py \
    [DATA_DIR]

# Multi-GPU Fully Sharded Data Parallel
torchrun \
    --nnodes 1 \
    --nproc_per_node=gpu    `# Number of processes per node, should be equal to the number of GPUs.` \
    train_fsdp.py \
    --batch-size 64         `# Batch size for training.` \
    --test-batch-size 1000  `# Batch size for testing.` \
    --epochs 10             `# Number of epochs to train.` \
    --lr 1.0                `# Learning rate.` \
    --gamma 0.7             `# Learning rate step gamma.` \

>     --save-model            `# Save the trained model.` \

Reviewer: ```suggestion
```
These aren't central to demonstrating Zeus, so I'd rather not have it here.
